pkg/cfn/manager/karpenter.go

```diff
@@ -0,0 +1,155 @@
+package manager
+
+import (
+	"context"
+	"fmt"
+
+	cfn "github.com/aws/aws-sdk-go/service/cloudformation"
+	"github.com/kris-nova/logger"
+	"github.com/pkg/errors"
+
+	api "github.com/weaveworks/eksctl/pkg/apis/eksctl.io/v1alpha5"
+	"github.com/weaveworks/eksctl/pkg/cfn/builder"
+	"github.com/weaveworks/eksctl/pkg/karpenter"
+	"github.com/weaveworks/eksctl/pkg/karpenter/providers/helm"
+)
+
+// KarpenterStack represents the Karpenter stack.
+type KarpenterStack struct {
+	KarpenterName string
+}
+
+// makeNodeGroupStackName generates the name of the Karpenter stack identified by its name, isolated by the cluster this StackCollection operates on
+func (c *StackCollection) makeKarpenterStackName() string {
+	return fmt.Sprintf("eksctl-%s-karpenter", c.spec.Metadata.Name)
+}
+
+// createKarpenterTask creates Karpenter
+func (c *StackCollection) createKarpenterTask(errs chan error) error {
+	name := c.makeKarpenterStackName()
+
+	logger.Info("building nodegroup stack %q", name)
+	stack := builder.NewKarpenterResourceSet(c.iamAPI, c.spec)
+	if err := stack.AddAllResources(); err != nil {
+		return err
+	}
+	tags := map[string]string{
+		api.KarpenterNameTag: name,
+	}
+	if err := c.CreateStack(name, stack, tags, nil, errs); err != nil {
+		return err
+	}
+	// Have to create these here, since the Helm Installer returns an error and I
+	// don't want to change StackCollection's New*. But this will make
+	// testing this function rather difficult.
+	helmInstaller, err := helm.NewInstaller(helm.Options{
+		Namespace: karpenter.KarpenterNamespace,
+	})
+	if err != nil {
+		return err
+	}
+	karpenterInstaller := karpenter.NewKarpenterInstaller(karpenter.Options{
+		HelmInstaller:         helmInstaller,
+		Namespace:             karpenter.KarpenterNamespace,
+		ClusterName:           c.spec.Metadata.Name,
+		AddDefaultProvisioner: false, // make this configurable
+		CreateServiceAccount:  true,  // make this configurable
+		ClusterEndpoint:       c.spec.Status.Endpoint,
+		Version:               c.spec.Karpenter.Version,
+	})
+	return karpenterInstaller.InstallKarpenter(context.Background())
+}
+
+// DescribeKarpenterStacks calls DescribeStacks and filters out karpenter resources
+func (c *StackCollection) DescribeKarpenterStacks() ([]*Stack, error) {
+	stacks, err := c.DescribeStacks()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(stacks) == 0 {
+		return nil, nil
+	}
+
+	var karpenterStacks []*Stack
+	for _, s := range stacks {
+		switch *s.StackStatus {
+		case cfn.StackStatusDeleteComplete:
+			continue
+		case cfn.StackStatusDeleteFailed:
+			logger.Warning("stack's status of karpenter named %s is %s", *s.StackName, *s.StackStatus)
+			continue
+		}
+		if c.GetKarpenterName(s) != "" {
```

Comment:
We should add a condition at the start of the loop instead that ignores non-Karpenter stacks.